storage/range.go

```diff
@@ -1677,20 +1733,66 @@ func (r *Range) Append(entries []raftpb.Entry) error {
 
 // ApplySnapshot implements the multiraft.WriteableGroupStorage interface.
 func (r *Range) ApplySnapshot(snap raftpb.Snapshot) error {
-	// TODO(bdarnell): discard any existing log that is obsoleted by this snapshot.
-	desc := &proto.RangeDescriptor{}
-	err := gogoproto.Unmarshal(snap.Data, desc)
+	snapData := proto.RaftSnapshotData{}
+	err := gogoproto.Unmarshal(snap.Data, &snapData)
 	if err != nil {
+		return nil
+	}
+
+	hardStateKey := engine.RaftHardStateKey(r.Desc().RaftID)
+	encodedHardStateKey := engine.MVCCEncodeKey(hardStateKey)
+	descKey := engine.RangeDescriptorKey(r.Desc().StartKey)
+	var desc proto.RangeDescriptor
+	seenDesc := false
+	batch := engine.NewBatch(r.rm.Engine())
+
+	// Delete everything in the range and recreate it from the snapshot.
+	for iter := newRangeDataIterator(r, r.rm.Engine()); iter.Valid(); iter.Next() {
+		if bytes.Equal(encodedHardStateKey, iter.Key()) {
```

Comment:
As mentioned previously, if we read the existing value first, allow this to write the entire batch of data, then rewrite previously existing hard state, we'll save a whole lot of CPU on comparing bytes.